presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeyValuePairs.java

```diff
@@ -103,4 +119,62 @@ public void add(Block key, Block value, int keyPosition, int valuePosition)
             }
         }
     }
+
+    private boolean keyExists(Block key, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        return keyPositionByHash[getHashPositionOfKey(key, position)] != EMPTY_SLOT;
+    }
+
+    private void addKey(Block key, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        int hashPosition = getHashPositionOfKey(key, position);
+        if (keyPositionByHash[hashPosition] == EMPTY_SLOT) {
+            keyPositionByHash[hashPosition] = keyBlockBuilder.getPositionCount() - 1;
```

Comment:
Would it make sense to put 

```
keyType.appendTo(key, keyPosition, keyBlockBuilder)
```

Right before this line? So `addKey` will add the key into the `keyBlockBuilder` and register it in the `keyPositionByHash`